File: cs/ch17-04-arithmetic/Money.Tests/CurrencyTransactionTest.cs

namespace TheSoftwareGorilla.TDD.Money.Tests;

#region TO DO's
#endregion
[TestFixture]
[Description("CurrencyTransactionTest class")]
public class CurrencyTransactionTest
{
    
    private Bank<Money> _bank;
    
    [SetUp]
    public void SetUp()
    {
        _bank = new Bank<Money>();
        _bank.AddRate("USD", "ZAR", 17.64m);
    }

    [TearDown]
    public void TearDown()
    {
        _bank = new Bank<Money>();
    }

    [TestCase]
    public void TestCurrencyTransaction()
    {
        var transaction = new CurrencyTransaction(Money.From(1000m, "USD", _bank), "ZAR");
        transaction.SourceFee = Money.From(35m, "USD", _bank);
        transaction.TargetCurrencyRateFeePercentage = 0.015m; 
        transaction.TargetServiceFee = Money.From(150m, "ZAR", _bank);
        transaction.Settle();
        Assert.That(transaction.TargetConversionRate, Is.EqualTo(17.3754m));
        Assert.That(transaction.TargetCurrencyFee.AmountIn("USD"), Is.EqualTo(15.00m));
        Assert.That(transaction.TargetCurrencyFee.AmountIn("ZAR"), Is.EqualTo(264.60m));
        Assert.That(transaction.TargetServiceFee.AmountIn("USD"), Is.EqualTo(8.50m));
        Assert.That(transaction.TargetServiceFee.AmountIn("ZAR"), Is.EqualTo(150.00m));
        Assert.That(transaction.TotalTargetFees.AmountIn("USD"), Is.EqualTo(23.50m));
        Assert.That(transaction.TotalTargetFees.AmountIn("ZAR"), Is.EqualTo(414.60m));
        Assert.That(transaction.SettlementAmount.AmountIn("USD"), Is.EqualTo(976.50m));
        Assert.That(transaction.SettlementAmount.AmountIn("ZAR"), Is.EqualTo(17225.40m));
        Assert.That(transaction.TotalTransactionFees.AmountIn("USD"), Is.EqualTo(58.50m));
        Assert.That(transaction.TotalTransactionFees.AmountIn("ZAR"), Is.EqualTo(1032.00m));
        Assert.That(transaction.TotalTransactionAmount.AmountIn("USD"), Is.EqualTo(1035.00m));
        Assert.That(transaction.TotalTransactionAmount.AmountIn("ZAR"), Is.EqualTo(18257.40m));
    }

    [TestCase]
    public void TestUnsettledTransaction()
    {
        var transaction = new CurrencyTransaction(Money.From(1000m, "USD", _bank), "ZAR");
        Assert.That(transaction.IsSettled, Is.EqualTo(false));
        Assert.That(transaction.SourceFee, Is.EqualTo( Money.From(0m, "USD", _bank)));
        Assert.That(transaction.TargetCurrencyRateFeePercentage, Is.EqualTo(0m)); 
        Assert.That(transaction.TargetServiceFee, Is.EqualTo( Money.From(0m, "ZAR", _bank)));
        Assert.That(transaction.TargetConversionRate, Is.EqualTo(0m));
        Assert.That(transaction.TargetCurrencyFee, Is.EqualTo( Money.From(0m, "ZAR", _bank)));
        Assert.That(transaction.TotalTargetFees, Is.EqualTo( Money.From(0m, "ZAR", _bank)));
        Assert.That(transaction.SettlementAmount, Is.EqualTo( Money.From(0m, "ZAR", _bank)));
        Assert.That(transaction.TotalTransactionFees, Is.EqualTo( Money.From(0m, "ZAR", _bank)));
        Assert.That(transaction.TotalTransactionAmount, Is.EqualTo( Money.From(0m, "USD", _bank)));
    }
}

File: java/ch17-04-arithmetic/src/test/java/com/thesoftwaregorilla/tdd/money/CurrencyTransactionTest.java
package com.thesoftwaregorilla.tdd.money;

import org.junit.jupiter.api.*;

import java.math.BigDecimal;

import static org.junit.jupiter.api.Assertions.*;

//<editor-fold desc="TO DO List">
//</editor-fold>

public class CurrencyTransactionTest {

    private static Bank<Money> bank;
    @BeforeAll
    public static void beforeAllSetUp() {
        bank = new Bank<>();
        bank.addRate("USD", "ZAR",new BigDecimal("17.64"));
    }

    @Test
    @DisplayName("currency transaction")
    public void testCurrencyTransaction() {
        CurrencyTransaction transaction = new CurrencyTransaction(Money.from(new BigDecimal("1000.00"), "USD", bank), "ZAR");
        transaction.setSourceFee(Money.from(new BigDecimal("35.00"), "USD", bank));
        transaction.setTargetCurrencyRateFeePercentage(new BigDecimal("0.015"));
        transaction.setTargetServiceFee(Money.from(new BigDecimal("150.00"), "ZAR", bank));
        transaction.settle();
        assertEquals(new BigDecimal("17.37540000"), transaction.getTargetConversionRate());
        assertEquals(new BigDecimal("15.00"), transaction.getTargetCurrencyFee().getAmountIn("USD"));
        assertEquals(new BigDecimal("264.60"), transaction.getTargetCurrencyFee().getAmountIn("ZAR"));
        assertEquals(new BigDecimal("8.50"), transaction.getTargetServiceFee().getAmountIn("USD"));
        assertEquals(new BigDecimal("150.00"), transaction.getTargetServiceFee().getAmountIn("ZAR"));
        assertEquals(new BigDecimal("23.50"), transaction.getTotalTargetFees().getAmountIn("USD"));
        assertEquals(new BigDecimal("414.60"), transaction.getTotalTargetFees().getAmountIn("ZAR"));
        assertEquals(new BigDecimal("976.50"), transaction.getSettlementAmount().getAmountIn("USD"));
        assertEquals(new BigDecimal("17225.40"), transaction.getSettlementAmount().getAmountIn("ZAR"));
        assertEquals(new BigDecimal("58.50"), transaction.getTotalTransactionFees().getAmountIn("USD"));
        assertEquals(new BigDecimal("1032.00"), transaction.getTotalTransactionFees().getAmountIn("ZAR"));
        assertEquals(new BigDecimal("1035.00"), transaction.getTotalTransactionAmount().getAmountIn("USD"));
        assertEquals(new BigDecimal("18257.40"), transaction.getTotalTransactionAmount().getAmountIn("ZAR"));
    }

    @Test
    @DisplayName("unsettled transaction")
    public void testUnsettledTransaction() {
        CurrencyTransaction transaction = new CurrencyTransaction(Money.from(new BigDecimal("1000.00"), "USD", bank), "ZAR");
        assertFalse(transaction.isSettled());
        assertEquals(Money.from(BigDecimal.ZERO, "USD", bank), transaction.getSourceFee());
        assertEquals(BigDecimal.ZERO, transaction.getTargetCurrencyRateFeePercentage());
        assertEquals(Money.from(BigDecimal.ZERO, "ZAR", bank), transaction.getTargetServiceFee());
        assertEquals(BigDecimal.ZERO, transaction.getTargetConversionRate());
        assertEquals(Money.from(BigDecimal.ZERO, "ZAR", bank), transaction.getTargetCurrencyFee());
        assertEquals(Money.from(BigDecimal.ZERO, "ZAR", bank), transaction.getTotalTargetFees());
        assertEquals(Money.from(BigDecimal.ZERO, "ZAR", bank), transaction.getSettlementAmount());
        assertEquals(Money.from(BigDecimal.ZERO, "ZAR", bank), transaction.getTotalTransactionFees());
        assertEquals(Money.from(BigDecimal.ZERO, "USD", bank), transaction.getTotalTransactionAmount());
    }

}
